### PR TITLE
feat[batch-submitter]: add lightweight metrics server

### PR DESCRIPTION
**Pull request by annieke**
**Description**
Added light server to batch submitter to serve `/metrics` endpoint for our dashboards.

**Additional context**

**Metadata**
- Fixes OP-701

### CODE REVIEW

File: packages/batch-submitter/src/exec/run-batch-submitter.ts
Comment (annieke) [inline packages/batch-submitter/src/exec/run-batch-submitter.ts:304]: @optimisticben @snario your feedback would be helpful here:

currently i'm only serving the transaction batch submitter's metrics since it's the one who care about most, but i'm wondering if the transaction batch submitter, state batch submitter, and batch submitter process should have separate `Metrics` instances at all. currently i'm initializing 3 of those classes in the file

https://github.com/ethereum-optimism/optimism/blob/ce7fa52b392f1932fc4f4ee10f4f22d23d9c0769/packages/batch-submitter/src/exec/run-batch-submitter.ts#L44-L47

https://github.com/ethereum-optimism/optimism/blob/ce7fa52b392f1932fc4f4ee10f4f22d23d9c0769/packages/batch-submitter/src/exec/run-batch-submitter.ts#L186-L189

https://github.com/ethereum-optimism/optimism/blob/ce7fa52b392f1932fc4f4ee10f4f22d23d9c0769/packages/batch-submitter/src/exec/run-batch-submitter.ts#L211-L214

is this best practice, or should they all share a registry?

File: packages/batch-submitter/src/exec/run-batch-submitter.ts
Comment (tynes) [inline packages/batch-submitter/src/exec/run-batch-submitter.ts:281]: What do you think of creating a class like so:

```js
class MetricsServer {
  constructor(options) {
    // assign particular options to this
    // validate options
  }
  async start() {}
  async stop() {}
}
```

Then that can be imported here and `start` can be called

Also keeping it lean is nice but without tests its hard to know if this will work. I wouldn't be opposed to using express since we use it in the DTL

Comment (annieke) [reply packages/batch-submitter/src/exec/run-batch-submitter.ts:281]: i think express would work! i'd like to leave this server easy to take apart/rebuild until our dashboards are up, since metrics as a whole is still taking form and this just serves that purpose

Comment (tynes) [reply packages/batch-submitter/src/exec/run-batch-submitter.ts:281]: > this just serves that purpose

With no tests there is no proof that the purpose is served?

I'd argue that having a simple class would make the server 10x easier to take apart and rebuild because to be honest I'd have to hack around a bit to figure out exactly what is going on with the low level `http` package in nodejs

File: packages/batch-submitter/src/exec/run-batch-submitter.ts
Comment (snario) [reply packages/batch-submitter/src/exec/run-batch-submitter.ts:304]: IMO Metrics registries should be tied to a _process_ and right now all three of these are in the same process.

Comment (optimisticben) [reply packages/batch-submitter/src/exec/run-batch-submitter.ts:304]: Agreed, sounds right to have a single metric registry.

File: packages/batch-submitter/src/exec/run-batch-submitter.ts
Comment (smartcontracts) [reply packages/batch-submitter/src/exec/run-batch-submitter.ts:281]: Yeah, I feel like this is potentially reusable code, IMO would be good to break this out into a small class and add a few basic tests.

Comment (snario) [reply packages/batch-submitter/src/exec/run-batch-submitter.ts:281]: Here is an example of a more functional approach: https://github.com/graphprotocol/common-ts/blob/5544327c0388a73f4e627625d077b64352031f9f/packages/common-ts/src/metrics/index.ts#L30

Comment (annieke) [reply packages/batch-submitter/src/exec/run-batch-submitter.ts:281]: thoughts on using the graph's implementation in this iteration?^ seems to serve our purpose

Comment (tynes) [reply packages/batch-submitter/src/exec/run-batch-submitter.ts:281]: We can use the graph's implementation for now, what do you think about putting it in `core-utils` and then it can be imported into each typescript service?

Comment (tynes) [reply packages/batch-submitter/src/exec/run-batch-submitter.ts:281]: Ideally we have a configurable http host/port as well as an flag to enable it optionally

Comment (annieke) [reply packages/batch-submitter/src/exec/run-batch-submitter.ts:309]: just updated this to use express and put it in common-ts

File: packages/common-ts/src/common/metrics.ts
Comment (tynes) [inline packages/common-ts/src/common/metrics.ts:57]: This should specify "metrics server is listening on port" - also specifying the hostname is important to be able to run this inside of a container

File: packages/batch-submitter/src/exec/run-batch-submitter.ts
Comment (tynes) [inline packages/batch-submitter/src/exec/run-batch-submitter.ts:311]: Can you put this behind an optional config option that defaults to `true`?

File: packages/batch-submitter/.env.example
Comment (snario) [inline packages/batch-submitter/.env.example:7]: `RUN_PROMETHEUS_SERVER`? We should also make the Prometheus port configurable

File: packages/batch-submitter/src/exec/run-batch-submitter.ts
Comment (snario) [inline packages/batch-submitter/src/exec/run-batch-submitter.ts:123]: True by default? Maybe it should depend on the `NODE_ENV`?

File: packages/batch-submitter/src/exec/run-batch-submitter.ts
Comment (snario) [inline packages/batch-submitter/src/exec/run-batch-submitter.ts:320]: We ought to have coerced this into a `boolean` by this point, I think

File: packages/batch-submitter/src/exec/run-batch-submitter.ts
Comment (snario) [inline packages/batch-submitter/src/exec/run-batch-submitter.ts:322]: Why store as a local variable, seems unused?

File: packages/common-ts/test/metrics.spec.ts
Comment (snario) [inline packages/common-ts/test/metrics.spec.ts:9]: ❤️

File: packages/batch-submitter/src/exec/run-batch-submitter.ts
Comment (tynes) [reply packages/batch-submitter/src/exec/run-batch-submitter.ts:123]: I don't think coupling it to `NODE_ENV` makes sense, it should be explicit

File: packages/batch-submitter/src/exec/run-batch-submitter.ts
Comment (tynes) [reply packages/batch-submitter/src/exec/run-batch-submitter.ts:322]: There is no top level class that it can be assigned to, ideally it would be:

```js
this.metricsServer = createMetricsServer({})
```

or better yet

```js
this.metricsServer = new MetricsServer({})
```

It shouldn't be garbage collected if its not assigned to a variable, the variable kind of does nothing but I suppose it looks kind of weird without it. This is the reason why I like the `start()` and `stop()` API that I suggested in https://github.com/ethereum-optimism/optimism/pull/792/files#r627706069

File: packages/common-ts/src/common/metrics.ts
Comment (tynes) [inline packages/common-ts/src/common/metrics.ts:58]: This logline should be something like: `Metrics server started`

Comment (tynes) [reply packages/common-ts/src/common/metrics.ts:58]: It should also log the route

File: packages/common-ts/src/common/metrics.ts
Comment (tynes) [inline packages/common-ts/src/common/metrics.ts:52]: Can you move the `||` outside of the `get()`?

**Review by smartcontracts — COMMENTED** [review]
Looking good. Will defer to @tynes and @snario on this one.

**Comment by annieke** [comment]
will make fixes after https://github.com/ethereum-optimism/optimism/pull/847 is merged!

File: packages/batch-submitter/src/exec/run-batch-submitter.ts
Comment (tynes) [inline packages/batch-submitter/src/exec/run-batch-submitter.ts:422]: hostname needs to be passed through here as well

Comment (tynes) [reply packages/batch-submitter/src/exec/run-batch-submitter.ts:422]: What do you think about making these options `metrics-server-*`? Just because I don't think they are technically coupled to prometheus, are they?

`hostname: config.str('metrics-server-hostname', '127.0.0.1')`

File: packages/common-ts/src/common/metrics.ts
Comment (tynes) [inline packages/common-ts/src/common/metrics.ts:57]: If there is a default port then we should have a default hostname as well

**Comment by tynes** [comment]
Needs a rebase and a couple of fixes then should be good to go

File: packages/batch-submitter/src/exec/run-batch-submitter.ts
Comment (annieke) [reply packages/batch-submitter/src/exec/run-batch-submitter.ts:422]: ^as of now they are! lol i switched because of this comment: https://github.com/ethereum-optimism/optimism/pull/792#discussion_r631310643 not opinionated about this but i think i'll stick with prometheus since our client is basically a wrapper

File: packages/batch-submitter/src/exec/run-batch-submitter.ts
Comment (tynes) [inline packages/batch-submitter/src/exec/run-batch-submitter.ts:462]: Personally I think that `if` statements that look like:

```
if (
  ...
) {
  ...
}
```

Are ugly and harder to read

**Review by tynes — APPROVED** [review]

File: packages/batch-submitter/src/exec/run-batch-submitter.ts
Comment (annieke) [reply packages/batch-submitter/src/exec/run-batch-submitter.ts:462]: linter automatically makes it that way because of the length :/

Comment (tynes) [reply packages/batch-submitter/src/exec/run-batch-submitter.ts:462]: Its possible to avoid that by doing a variable assignment 1 line above and it ends up making it take up less lines of code

**Comment by tynes** [comment]
Will merge once tests are passing

**Pull request merged**
